clippy_lints/src/ptr.rs

```diff
@@ -250,13 +251,22 @@ impl<'tcx> LateLintPass<'tcx> for Ptr {
     }
 
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
-        if let ExprKind::Binary(ref op, l, r) = expr.kind {
-            if (op.node == BinOpKind::Eq || op.node == BinOpKind::Ne) && (is_null_path(cx, l) || is_null_path(cx, r)) {
-                span_lint(
+        if let ExprKind::Binary(op, l, r) = expr.kind {
+            if op.node == BinOpKind::Eq || op.node == BinOpKind::Ne {
+                let non_null_path_snippet = match (is_null_path(cx, l), is_null_path(cx, r)) {
+                    (true, false) if let Some(sugg) = Sugg::hir_opt(cx, r) => sugg.maybe_par().to_string(),
+                    (false, true) if let Some(sugg) = Sugg::hir_opt(cx, l) => sugg.maybe_par().to_string(),
+                    _ => return,
+                };
```

Comment:
We can drop a `String` allocation here by doing something like
```suggestion
            if op.node == BinOpKind::Eq || op.node == BinOpKind::Ne 
                && let Some(non_null_path_snippet) = match (is_null_path(cx, l), is_null_path(cx, r)) {
                    (true, false) => Sugg::hir_opt(cx, r),
                    (false, true) => Sugg::hir_opt(cx, l),
                    _ => return,
                }
            {
```

And `.maybe_par()` in the `format!`

Reply:
Thanks, fixed.